### ISSUE TITLE
slow performance on google page speed due to zone.js

### ISSUE THREAD

**Issue by jcdsr** [angular/angular#32103]
I Use angular7 as ssr, and we found out that our application has heavy and serious problems with google page speed and all the issues are funnelling to zone.js

Page speed complain about the script valuation is too high, after breaking down the app, we found the issues were related to polyfills, however, polyfills only containing

```
  import 'zone.js/dist/zone.js';  // Included with Angular CLI.
  import 'classlist.js/classList.min.js';
  import 'web-animations-js/web-animations.min.js';
```

then what we did was totally remove all the content on polyfills file, leave it empty, and add those files, directly from a cdn on the bottom of index.html as:

```
<script src="https://cdnjs.cloudflare.com/ajax/libs/zone.js/0.9.1/zone.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/classlist/1.2.201711092/classList.min.js"></script>
```

and the result was the same, the page speed still complaining about script valuation, however, we manage to reduce 50% of the script valuation, however, still and is really bad.

we were using a zone.js@0.8.10 then we upgrade to latest but nothing. Same result.

In fact, this is affecting all our apps no matter if we build as aot or server-side render, the result is the same.

Does anyone know how to solve this issue with page speed? Our ranking is below 20 over 100 as SSR.

This doesn't make any sense, it supposes to be faster, not slower?
I also try
`<script src="https://cdnjs.cloudflare.com/ajax/libs/zone.js/0.9.1/zone-evergreen.min.js"></script>`

---

I found plenty of sites with the same issue, related to polyfills.ts or have problems with google page speed, you can see here https://www.madewithangular.com/categories/angular/ that any site made with angular have this issue with zone.js Please check any site, copy and paste the URL into page speed and you will see the same issue, over and over again, this is not a coincidence.

https://www.delta.com/eu/en
https://www.forbes.com

and many more have... Angular and zone.js must address this issue asap!

This affects all the angular SPA websites or Angular Universal or SSR

**Comment by Gnorro** [comment]
This is a very interesting question We have exactly the same problem.

We already made lots of changes in our application (lazy loading, reduced a lot main bundle and so on...). We have SSR too, but Google Pagespeed Insight give always a bad result (from 0 to 20 max).

First Contentful Paint is good (from 1.2 to 2.5 seconds), but what is bad is TTI and script evaluation.

I think SPA will always suffer of long script evaluation and will always have a worse score compared to static pages.

I don't know if there is a solution. I hope I'm wrong

**Comment by jcdsr** [comment]
I found plenty of sites with the same issue, related to polyfills.ts or have problems with google page speed, you can see here https://www.madewithangular.com/categories/angular/ that any site made with angular have this issue with zone.js Please check any site, copy and paste the URL into page speed and you will see the same issue, over and over again, this is not a coincidence.

https://www.delta.com/eu/en
https://www.forbes.com

and many more have... Angular and zone.js must address this issue asap!

**Comment by Mylogin-info** [comment]
We have the same issue. Occured around update from 7.1 to 7.2. We had to change the "MediaService" to "MediaObserver". And the "ViewChild" _(You must now specify that change detection should run before results are set. )_

We are now at Angular 9 and "zone.js": "^0.10.2" and also tried lower versions.

Resizing was no problem before 7.1. Now it's very slow/sluggish. Takes up to 4 seconds.

Performance warning:
- Task - Long task took 4.28 s.
- Event: resize - Handler took 162.42 ms
- Recalculate Style - Forced reflow is a likely performance bottleneck.

Here is the same app: _(resize both to trigger responsiveness and see the performance difference)_

**Not updated:**
https://secure.mylogin.info/reports/big-five/1.0/?id=bigfive.json

**Updated:**
https://secure.mylogin.info/reports/big-five-performance-demo/1.0/?id=bigfive.json

I hope this is helpfull. I tried to make a stackblitz demo, but was unsuccessful.

**Comment by Gnorro** [comment]
It's frustrating to know that even if you follow all the best practises and performance guides around  the world, it will not be enough to have a decent score. I'm not talking fo100%, I mean something bigger than 20-30%...

I think google devs should focus on this because it's a common problem and have a very big impact on SEO. They suggest us to use Angular but than they penalyze these apps..,

It doesn't make sense

**Comment by jcdsr** [comment]
Not just we have issues with angular and Zone.js but lighthouse reports are wrong, check this: https://github.com/GoogleChrome/lighthouse/issues/6708

**Comment by Gnorro** [comment]
Lighthouse is a bad tool to measure performance

if you run it on 2 different PC, one with recent hardware, processor etc... and one less powerfull and run lighthouse you will get 2 different results. I'm not talking about having some points of difference, I mean having a difference of 50%

That tool is deeply related to performance of the machine where is running.

You should use some external tool, for example https://developers.google.com/speed/pagespeed/insights/

**Comment by JiaLiPassion** [comment]
I will check this issue, thanks!

**Comment by Mylogin-info** [comment]
@JiaLiPassion That's great! If you need more info, please let me know.

**Comment by jcdsr** [comment]
Any ideas?

**Comment by JiaLiPassion** [comment]
Sorry for the late reply.
@Mylogin-info, if `resize` is slow, and you don't have any resize event listener to update component data, you can just make `resize` outside of zone by define a `variable` called `__zone_symbol__UNPATCHED_EVENTS = ['resize'];`

If you have any other performance issue, could you provide a reproduce repo?

@jcdsr, could you provide some repo to reproduce, thanks!

**Comment by Mylogin-info** [comment]
@JiaLiPassion Thank you! I wil try this and let you know what the result is.

**Comment by jcdsr** [comment]
@JiaLiPassion 
That works, thanks. The script valuation is now around 1,061 ms

However, I couldn't find any documentation regarding these zone symbols. How did you find out that was specifically 'resize', what if there is another one? how we can debug that? also, my app is SSR as angular 8. 

I add the code on polyfills like:

is this correct?

`
        if (typeof window !== "undefined") {
            /* https://github.com/angular/angular/issues/32103 */
            (window as any).__zone_symbol__UNPATCHED_EVENTS = ['resize'];
        }
        import 'zone.js';
`

**Comment by rkajbaf** [comment]
In the default polyfill.ts auto-generated by the CLI (when you create a new project using the CLI) has this in the comments right above the `import 'zone.js/dist/zone'`  line you fill find this:

```
/**
 * By default, zone.js will patch all possible macroTask and DomEvents
 * user can disable parts of macroTask/DomEvents patch by setting following flags
 * because those flags need to be set before `zone.js` being loaded, and webpack
 * will put import in the top of bundle, so user need to create a separate file
 * in this directory (for example: zone-flags.ts), and put the following flags
 * into that file, and then add the following code before importing zone.js.
 * import './zone-flags.ts';
 *
 * The flags allowed in zone-flags.ts are listed here.
 *
 * The following flags will work for all browsers.
 *
 * (window as any).__Zone_disable_requestAnimationFrame = true; // disable patch requestAnimationFrame
 * (window as any).__Zone_disable_on_property = true; // disable patch onProperty such as onclick
 * (window as any).__zone_symbol__UNPATCHED_EVENTS = ['scroll', 'mousemove']; // disable patch specified eventNames
 *
 *  in IE/Edge developer tools, the addEventListener will also be wrapped by zone.js
 *  with the following flag, it will bypass `zone.js` patch for IE/Edge
 *
 *  (window as any).__Zone_enable_cross_context_check = true;
 *
 */
```

That's where it comes from.  So you would just put 'resize' to the list for `__zone_symbol__UNPATCHED_EVENTS `

**Comment by viters** [comment]
Are there any more problems around zone and angular startup performance?
I have similar situation to @Gnorro 

> We already made lots of changes in our application (lazy loading, reduced a lot main bundle and so on...). We have SSR too, but Google Pagespeed Insight give always a bad result (from 0 to 20 max).

Sadly, setting __zone_symbol__UNPATCHED_EVENTS didn't really help. I have no idea where to look for optimizations. PageSpeed shows main thread 12s and js execution 8s, where most of it is angular startup - which decreases my page score a lot.

**Comment by jcdsr** [comment]
Seems not working by adding 

    (window as any).__Zone_disable_requestAnimationFrame = true; 
    (window as any).__zone_symbol__UNPATCHED_EVENTS = ['resize']; 

![image](https://user-images.githubusercontent.com/4386097/72537631-5e704800-3874-11ea-9827-b056ae5f76c3.png)

Here a link of the polyfills after build
https://github.com/jcdsr/polyfill-test-1.js

**Comment by JiaLiPassion** [comment]
@jcdsr , please provide a workable sample, and the screenshot can not explain which part is slow.

**Comment by jcdsr** [comment]
Sorry but I'm not allowed to share the project.

**Comment by mlc-mlapis** [comment]
@jcdsr Nobody wants you to share the production code. Just abstracted code which is simple enough, but still working and demonstrating the problem. How do you imagine that anybody can do anything if there is no code to understand?

**Comment by Gnorro** [comment]
mlc-mlapis consider that maybe the problem is not so clear so it's not easy to understand how to reproduce the problem.

For example we have a big app and we have performance problem but we don't know where exactly. How do you expect we can post some code removing parts of the app?

I understand also that is not easy to try to resolve these problem without code.

**Comment by mlc-mlapis** [comment]
@Gnorro You would seriously think about the possibility to tightly cooperate with somebody from Angular core team or somebody who can help you like @JiaLiPassion. I understand that your app is not a public one, but in the case that your app is not the top-secret thing (indeed not all the parts), you should find a way how to allow access and diagnostic it to them.

**Comment by jcdsr** [comment]
What exactly you need? because I sent a link previously with a copy of the polyfills after build and in that link shows what do I use on that file. I'm not sure what you looking

**Comment by JiaLiPassion** [comment]
@jcdsr, I need the source repo and a reproduce environment.

**Comment by jcdsr** [comment]
@JiaLiPassion 
on https://github.com/angular/zone.js/issues/691
You comment we should disable long-stack-trace on production

How we can tell is totally disabled on Angular 8?

**Comment by kirillgroshkov** [comment]
Seems like we have a similar problem in our app. But the app is rather big and I don't know how to produce "minimal sample". App is public on the internet and source can be provided on-demand. Can someone help? We can pay for such work.

**Comment by unrealisted** [comment]
Any updates? I have the same issue. Thanks

**Comment by sudhakarverma610** [comment]
i am also having same problem with zone js when i remove then page speed goes till 96%+ when i put back then gtmetrix score not going above to 60% please help me ASAP

**Comment by sudhakarverma610** [comment]
i have same problem due to zone js i have implemented lazy loading and defer but i am unable to resolve total CPU time  and Script Evaluation >2 second   which lead bad performance please Tell me How can i Resolve this ASAP

**Comment by JiaLiPassion** [comment]
Please provide reproduce repo so I can help to do some research.

**Comment by jcdsr** [comment]
The issue is Zone.js - Zone.js is slowing down the application on google page speed, I have this problem (for years). You can try too by disabling the zone.js on polyfills, and then add the zone.js file on assets and load as a script tag on the index.html and you will see the issue, and then disable the zone.js tag and run the lighthouse.

Test google page speed: 

 1. normal zone.js in polyfills. 
 2. Zone.js in assets with script tag on the index.html (on the head) - make sure to disable on polyfills
 3. Disable zone.js on disable on polyfills and or index.html

**Comment by jcdsr** [comment]
I even try to load zone.js when is need as this post
https://stackoverflow.com/questions/45754526/how-to-only-load-zone-js-when-it-is-needed

But this didn't work either. Zone.js is slowing down all angular apps, no matter what version is

**Comment by jcdsr** [comment]
See the result on my side as angular universal, express.js and serving as pm2

With zone.js on polyfills.ts 

![Screenshot 2021-01-28 at 10 32 12](https://user-images.githubusercontent.com/4386097/106126048-cda19d00-6154-11eb-8d3a-d27ee32d3677.png)

![Screenshot 2021-01-28 at 10 30 45](https://user-images.githubusercontent.com/4386097/106126063-d09c8d80-6154-11eb-9687-7fe54190f096.png)


Disabling zone.js and adding to assets and loading as script tag on index.html plus disabling zone.js in polyfills


![Screenshot 2021-01-28 at 10 34 33](https://user-images.githubusercontent.com/4386097/106126147-ed38c580-6154-11eb-8476-32ef6ca20a6e.png)
 
![Screenshot 2021-01-28 at 10 34 47](https://user-images.githubusercontent.com/4386097/106126155-ef9b1f80-6154-11eb-8822-54e79d6a6bc3.png)

Disabling zone.js on index.html and polyfill - 

![Screenshot 2021-01-28 at 10 35 46](https://user-images.githubusercontent.com/4386097/106126228-080b3a00-6155-11eb-8179-c3f7d7afbc87.png)

![Screenshot 2021-01-28 at 10 35 59](https://user-images.githubusercontent.com/4386097/106126235-09d4fd80-6155-11eb-98f3-bfaf2240ed0d.png)

**Comment by JiaLiPassion** [comment]
@jcdsr, since `zone.js` patch all async APIs, so if you load zone.js, all time of `event listeners`, `setTimeout`, `Promise` will be counted on `zone.js`, so I still need a reproduce repo to find out what is the reason to cause your app slow with `zone.js`.

**Comment by Gnorro** [comment]
This is a chronic problem of angular with medium/large apps.

There is no solution. Angular app will be always slower than classic webapp
(the hope is that in the future google devs will change something).

A small solution is to develop entire app loading also components in lazy,
not only modules. This will make the main bundle smaller, but it will not
solve the problem.

The other solution (that is the one we choosed) is to have a service like
seo4ajax, prerender.io and similars.

I'm sorry to give you this bad news, but we have been working on this for
years, without a definitive solution

Bye


Il giorno ven 29 gen 2021 alle ore 09:01 JiaLiPassion <
notifications@github.com> ha scritto:

**Comment by mlc-mlapis** [comment]
@jcdsr You have 14 / 16 / 10 secs for Time to Interactive, so there have to be also other factors than Zone for sure. There is the only way. You have to reduce the application, step by step, to a minimal form to reproduce the hidden problem.

**Comment by sudhakarverma610** [comment]
i can't shared repo to  @JiaLiPassion because repo is private which is shared to me by client  
but i saw one thing that in gtmertix when i am using firefox in mumbai india then it's showing me very well report 
![image](https://user-images.githubusercontent.com/44521928/106270273-b9909500-6253-11eb-980d-94bd6bc1fe87.png)
but when i tested in chrome its showing me very bad report 
![image](https://user-images.githubusercontent.com/44521928/106270397-f78db900-6253-11eb-9eb5-fc5090251ed8.png)
Can you please provide information why this is happening.
i also  saw when i try to Open in https://y.jewelry/en-in/ its load quickly but still performance bad but when i try to open sub module url(https://y.jewelry/en-in/info/about) in gtmetrix then gtmetrix get me error timeout because it will taking so much time. so can you please provide me relevant solution

**Comment by BruneXX** [comment]
It seems that if @jcdsr doesn't share a repo @JiaLiPassion will not go deeper on this, please is there a way someone can share a repo to reproduce this issue? I'll try but I don't know when I'll have it ready, also I think this is easily reproducible in any angular app, SSR or not.

**Comment by sertsedat** [comment]
we just tried to create a new angular app (via cli) and it had a really bad lighthouse score as well.

**Comment by NeerajThapliyal** [comment]
i have removed all routes but still Script Evolution taking 1700 ms. 

can anyone help me on this?
![image](https://user-images.githubusercontent.com/29271119/121802342-64c94600-cc59-11eb-9b7c-2497718941b5.png)
![image](https://user-images.githubusercontent.com/29271119/121802371-8a564f80-cc59-11eb-9504-510636178156.png)

**Comment by jonathasborges1** [comment]
#up

**Comment by JiaLiPassion** [comment]
@davidstellini , I checked your repo, it seems it is using `animation` a lot, could you try this?

1. create a `zone-flags.ts` file under `src`.
2. add `(window as any).__Zone_disable_requestAnimationFrame = true;` into `zone-flags.ts`.
3. update `src/polyfills.ts`
```
import './zone-flags'; // add this line
import 'zone.js/dist/zone'; // Included with Angular CLI.
```

**Comment by hiepxanh** [comment]
I have idea, what if we lazyload zone? can we do it? I can pull zone once the user move the mouse or start scroll? is that legal @JiaLiPassion ?

**Comment by hiepxanh** [comment]
![image](https://user-images.githubusercontent.com/16970990/157583831-feb50d3b-c445-422d-ad52-e1783573c678.png)

wew, it have good improve total blocking time for me

**Comment by sshanmu91** [comment]
Guys! Any Update on the Slowness Issue!
With Angular v10 & zone.js of v0.11.5,
On Few of the Clicks on the Tabs, The Route Navigation, (OR) Sometimes Few of the Dropdown Clicks also,
We experience the Slowness in the Page Navigation or in Opening the Dropdown Options List.

With DynaTrace Enabled, When the Real-User-Montinoring Agent gets injected,
Their Event to Capture the Actions on Click of Certain Elements in the Application,
We experience the Slowness.


![image](https://user-images.githubusercontent.com/36123879/171199616-3dbe48c1-5f5c-4d12-a37d-c1ce3da5b967.png)


![image](https://user-images.githubusercontent.com/36123879/171199971-15dcea02-31bc-43a0-a474-02f2710e3273.png)


![image](https://user-images.githubusercontent.com/36123879/171200085-e0b9f0c6-b488-4fb3-a2d5-9b9d378484b6.png)


Could you please help us on this!

**Comment by BruneXX** [comment]
IMHO, I was using angular universal in a proyect for a long time and I finally decied to go with React, I think angular is not yet prepared for SSR as well as React, so I decided to switch. Also documentation is not enough to go with angular is hard to start with universal for those that are coming from non SSR apps and diffcult to optimize both worlds, server side and client side, too much work for these kind of results...

**Comment by maxisam** [comment]
@BruneXX Technically React doesn't handle SSR. Nextjs and other framework take care of that. It is ashamed that Angular haven't put enough resource in this area. The good news is it seems like it will be from the Angular blog 

https://blog.angular.io/angulars-vision-for-the-future-3cfca5e7b448

**Comment by hisham** [comment]
I personally use https://scully.io/ pre-rendering with Angular for our websites. That setup has worked with us, but Scully is relatively early. Not sure how it compares with NextJS as I never tried it. Angular anyways has plans to make Zone.js optional: https://angular.io/guide/roadmap

**Comment by jessicajaniuk** [comment]
There isn't much actionable here and we're going to close this as it seems like more of a support issue related to application structuring. There are limitations to zonejs in terms of performance, and while there is a future where zonejs is optional, until then there's nothing actionable here.

**Comment by prasannaatibc** [comment]
Hello!
I'm running an Angular project with version listed below

Angular CLI: 13.2.6
Node: 14.15.0
Package Manager: npm 6.14.8
OS: linux x64

Angular: 13.2.7
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, platform-server
... router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1302.6
@angular-devkit/build-angular   13.2.6
@angular-devkit/core            13.2.6
@angular-devkit/schematics      13.2.6
@angular/cdk                    13.3.1
@angular/cli                    13.2.6
@angular/material               13.3.1
@angular/pwa                    14.0.1
@angular/service-worker         13.1.1
@nguniversal/builders           13.1.1
@nguniversal/express-engine     13.0.2
@schematics/angular             14.0.1
rxjs                            7.5.5
typescript                      4.5.5

I have checked my application's performance through google pagespeed insights. The result stated that the Total Blocking Time(TBT) as 400ms which is not a good one. Below is the screenshot of the result.
![image](https://user-images.githubusercontent.com/92422921/183436520-dd0654bb-cc64-48d1-95dd-616e7197d7f5.png)

If you notice the polyfills file is the reason behind it. I have imported only "zone.js" inside the polyfills file(Nothing more than that). I have tried removing the "zone.js" import in the polyfills file. The performance score went 90 and 90+ but my application's events were not working.

To be noted, my browser console shows the below error.
![image](https://user-images.githubusercontent.com/92422921/183442476-d48b9491-f2d1-4829-9d77-e967f182d1ed.png)

As suggested by other developers on different blogs I have tried the following steps to reduce the TBT caused by polyfills file.
1. Commented out zone.js import in polyfills.ts and imported in index.html through script tag.
2. Add platformBrowserDynamic().bootstrapModule(AppModule, {ngZone: "noop"}) in main.ts
3. Imported polyfills file using javascript's document.createElement() function

But nothing helped. Can anyone give me a solution for reducing TBT caused by polyfills file and to improve my application's performance score in  pagespeed insights. Thanks in advance!
